Core/Quests: automatically add quests with AUTO_PUSH flag to player

matanshukry opened the pull request:
**Changes proposed:**

-  automatically push quests with relevant flag on player login

**Target branch(es):** 3.3.5/master

- [ ] 3.3.5
- [x] master

**Issues addressed:**

**Tests performed:**

(Does it build, tested in-game, etc.)

- [x] It builds
- [x] Tests in game

**Known issues and TODO list:** (add/remove lines as needed)

- [ ] Currently a bug: need to filter the quests and somehow and not push them all. Possibly by zone/area id.

matanshukry committed:
Calculating which quests should be sent to client on startup as they are
auto-push. 1st step as it's missing packet info

Aokromes commented on src/server/game/Server/Packets/QuestPackets.cpp:
```diff
@@ -746,5 +746,6 @@ void ChoiceResponse::Read()
     _worldPacket >> ChoiceID;
     _worldPacket >> ResponseID;
 }
+
```
unneeded change

Aokromes commented on src/server/game/Handlers/CharacterHandler.cpp:
```diff
@@ -1365,6 +1365,8 @@ void WorldSession::HandlePlayerLogin(LoginQueryHolder* holder)
     sScriptMgr->OnPlayerLogin(pCurrChar, firstLogin);
 
+
+
```
unneeded change.

matanshukry committed:
Simply adding the relevant quest

illfated commented:
PR title suggestion: **Core/Quests: push quests automatically with AUTO_PUSH flag**

matanshukry committed:
Removed extra unncessary lines

matanshukry commented:
Thanks @mdX7 for helping me figuring out how to push it to the player!

which looks a lot simpler now that we know what the solution is

matanshukry committed:
Added more quest sort enums from dbc

matanshukry committed:
Merge branch 'issue/add-quest-sort' into issue/auto-push-quests

matanshukry committed:
Merge remote-tracking branch 'origin/master' into issue/auto-push-quests

Shauren commented on src/server/game/Quests/QuestDef.h:
```diff
@@ -531,6 +531,8 @@ class TC_GAME_API Quest
         uint32 CalculateHonorGain(uint8 level) const;
         bool   CanIncreaseRewardedQuestCounters() const;
 
+        bool IsAutoPush() const { return HasFlagEx(QuestFlagsEx::QUEST_FLAGS_EX_AUTO_PUSH); }
```
`QuestFlagsEx` is a regular enum, not enum class, so don't add `QuestFlagsEx::`

Shauren commented on src/server/game/Entities/Player/Player.cpp:
```diff
@@ -18742,9 +18742,34 @@ bool Player::LoadFromDB(ObjectGuid guid, CharacterDatabaseQueryHolder* holder)
+
+    for (ObjectMgr::QuestMap::const_iterator it = questMap.cbegin(); it != questMap.cend(); ++it)
+    {
+        QuestStatusMap::const_iterator result = m_QuestStatus.find(it->first);
```
All of the checks (including quest status) should be replaced with just a simple function call `CanTakeQuest(it->second, false)`

Shauren commented on src/server/game/Globals/ObjectMgr.h:
```diff
@@ -1701,6 +1702,7 @@ class TC_GAME_API ObjectMgr
         std::map<HighGuid, std::unique_ptr<ObjectGuidGeneratorBase>> _guidGenerators;
 
         QuestMap _questTemplates;
+        QuestMap _questTemplatesAutoPush;
```
You can just turn this into a `std::vector<Quest const*>`

Shauren commented on src/server/game/Entities/Player/Player.h:
```diff
@@ -2574,6 +2574,8 @@ class TC_GAME_API Player : public Unit, public GridObject<Player>
         /***                    QUEST SYSTEM                   ***/
         /*********************************************************/
 
+        void _PushQuests();
```
remove `_` from function name

matanshukry replied:
sure. I noticed other quests that are starting with _, I thought it's a notion for private methods.
When should one use _ for method names then?

illfated replied:
Usually it is reserved for class member variables, but looking at the start of this file, I presume there is an exception of some description for the utterly few methods starting with underscore.

Shauren commented on src/server/game/Entities/Player/Player.cpp:
```diff
@@ -18742,9 +18742,34 @@ bool Player::LoadFromDB(ObjectGuid guid, CharacterDatabaseQueryHolder* holder)
     m_achievementMgr->CheckAllAchievementCriteria(this);
     m_questObjectiveCriteriaMgr->CheckAllQuestObjectiveCriteria(this);
+
+    _PushQuests();
```
Any other cases when quests are pushed to players? Levelup? Zone change?

matanshukry replied:
I added zone change before I saw this, I forgoet levelup though. I'll do that too.

matanshukry committed:
- Removed check for world quests on criteria tree complete
- Instead of pushing all wqs, now only pushing based on zone + not wqs + not deprecated, and verifying we can take it.

matanshukry committed:
- Renamed _PushQuests to PushQuests
- Changed questTemplatesAutoPush from map to vector (not querying by id)
- Removed enum prefix for QuestFlags and QuestFlagsEx (they're not enum class)

matanshukry committed:
Cast zoneOrMapId to uint32 before comparison

illfated commented:
Good job so far.

matanshukry committed:
Merge branch 'master' into issue/auto-push-quests

matanshukry committed:
Merge remote-tracking branch 'origin/master' into issue/auto-push-quests

Shauren commented on src/server/game/Entities/Player/Player.cpp:
```diff
@@ -18756,9 +18759,37 @@ bool Player::LoadFromDB(ObjectGuid guid, CharacterDatabaseQueryHolder* holder)
+
+    for (Quest const *quest : questsAutoPush)
+    {
+        QuestStatusMap::const_iterator result = m_QuestStatus.find(quest->GetQuestId());
```
Don't check `m_QuestStatus` here, `CanTakeQuest` already does that for you

Shauren reviewed (changes requested):
I have investigated the quest list that have this flag and here is what I found:
* All of them are candidates for pushing, regardless of the zone you are in (remove that condition)
* Almost all of them are quest types currently unsupported (emissary quests)
* The condition to filter out unsupported quests is `!quest->GetQuestTag() || quest->GetQuestTag() == QuestTagType::Tag` (GetQuestTag() returns Optional)

matanshukry committed:
Only checking quest tag now rather than zone

Shauren commented on src/server/game/Entities/Player/Player.cpp:
```diff
@@ -18756,9 +18759,26 @@ bool Player::LoadFromDB(ObjectGuid guid, CharacterDatabaseQueryHolder* holder)
+            continue;
+
+        if (!quest->IsUnavailable() && CanTakeQuest(quest, false))
+            AddQuest(quest, nullptr);
```
Shouldn't this be `AddQuestAndCheckCompletion(quest, nullptr);`?

matanshukry replied:
Usually I would say no, but I noticed the SL quest progress is kept (both tc and on retail), so yes. Either way it's probably safer to use it than not.

matanshukry committed:
Also checking completion when adding a quest.

Shauren commented on src/server/game/Entities/Player/Player.cpp:
```diff
@@ -18756,9 +18759,26 @@ bool Player::LoadFromDB(ObjectGuid guid, CharacterDatabaseQueryHolder* holder)
     return true;
 }
 
+void Player::PushQuests()
+{
+    std::vector<Quest const*> const& questsAutoPush = sObjectMgr->GetQuestTemplatesAutoPush();
+
+    for (Quest const *quest : questsAutoPush)
+    {
+        //
+        if (quest->GetQuestTag() && quest->GetQuestTag() != QuestTagType::Tag)
+            continue;
+
+        if (!quest->IsUnavailable() && CanTakeQuest(quest, false))
+            AddQuestAndCheckCompletion(quest, nullptr);
+    }
+}
+
```
```suggestion
void Player::PushQuests()
{
    for (Quest const* quest : sObjectMgr->GetQuestTemplatesAutoPush())
    {
        if (quest->GetQuestTag() && quest->GetQuestTag() != QuestTagType::Tag)
            continue;

        if (!quest->IsUnavailable() && CanTakeQuest(quest, false))
            AddQuestAndCheckCompletion(quest, nullptr);
    }
}

```

Shauren commented on src/server/game/Globals/ObjectMgr.cpp:
```diff
@@ -4214,6 +4215,9 @@ void ObjectMgr::LoadQuests()
 
         Quest* newQuest = new Quest(fields);
         _questTemplates[newQuest->GetQuestId()] = newQuest;
+
+        if (newQuest->IsAutoPush())
+            _questTemplatesAutoPush.push_back(newQuest);
```
```suggestion
        if (newQuest->IsAutoPush())
            _questTemplatesAutoPush.push_back(newQuest);
```

Shauren committed:
Apply suggestions from code review

Shauren commented on src/server/game/Entities/Player/Player.cpp:
```diff
@@ -18756,9 +18759,26 @@ bool Player::LoadFromDB(ObjectGuid guid, CharacterDatabaseQueryHolder* holder)
     return true;
 }
 
+void Player::PushQuests()
+{
+    std::vector<Quest const*> const& questsAutoPush = sObjectMgr->GetQuestTemplatesAutoPush();
+
+    for (Quest const *quest : questsAutoPush)
+    {
+        //
```
```suggestion
void Player::PushQuests()
{
    for (Quest const* quest : sObjectMgr->GetQuestTemplatesAutoPush())
    {
```

Shauren committed:
codestyle
